tests/gen/gen_auto.go

```diff
@@ -0,0 +1,245 @@
+	if config == nil {
+		config = make(configsMap)
+	}
+	configurator := newDocGenConfigurator(types, config)
+	return &randomDocGenerator{
+		configurator:     configurator,
+		generatedDocKeys: make(map[string][]string),
+	}
+}
+
+type randomDocGenerator struct {
+	configurator docsGenConfigurator
+
+	generatedDocKeys map[string][]string
+	random           rand.Rand
+}
+
+func (g *randomDocGenerator) generateDocs(options ...Option) ([]GeneratedDoc, error) {
+	err := g.configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *g.configurator.random
+
+	resultDocs := make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(g.configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.configurator.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			resultDocs = append(resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.configurator.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.configurator.usageCounter.getNextTypeIndForField(secondaryType, field)
+	docKey := g.generatedDocKeys[field.Schema][ind]
+	return docKey
+}
+
+func (g *randomDocGenerator) getDocKey(typeDef *client.CollectionDefinition, doc map[string]any) (string, error) {
+	clientDoc, err := client.NewDocFromJSON([]byte(createDocJSON(typeDef, doc)))
+	if err != nil {
+		return "", err
+	}
+	return clientDoc.Key().String(), nil
+}
+
+func (g *randomDocGenerator) generateRandomDocs(order []string) ([]DocsList, error) {
+	result := []DocsList{}
+	for _, typeName := range order {
+		col := DocsList{ColName: typeName}
+		typeDef := g.configurator.types[typeName]
+
+		currentTypeDemand := g.configurator.docsDemand[typeName]
+		averageDemand := currentTypeDemand.getAverage()
+		for i := 0; i < averageDemand; i++ {
+			newDoc := make(map[string]any)
+			for _, field := range typeDef.Schema.Fields {
+				if field.Name == request.KeyFieldName {
+					continue
+				}
+				if field.IsRelation() {
+					if field.IsPrimaryRelation() {
+						newDoc[field.Name+request.RelatedObjectID] = g.getNextPrimaryDocKey(typeName, &field)
+					}
+				} else {
+					fieldConf := g.configurator.config.ForField(typeName, field.Name)
+					newDoc[field.Name] = g.generateRandomValue(typeName, field.Kind, fieldConf)
+				}
+			}
+			docKey, err := g.getDocKey(&typeDef, newDoc)
+			if err != nil {
+				return nil, err
+			}
+			g.generatedDocKeys[typeName] = append(g.generatedDocKeys[typeName], docKey)
```

Comment:
suggestion: A code comment here explaining why you are appending the dockey here may save some future head scratching - it is not obvious glancing over the code, especially given that the map is not read from in this function.  I'd recommend highlighting why it needs to be held on the host object and can't be a local var passed into `getNextPrimaryDocKey` too, as that is also not obvious (and I assume that it is needed, if not please remove from the host state).

Reply:
refactored